### PR TITLE
Adaptive srk

### PR DESCRIPTION
**Pull request by frankschae**
Continues PR #331 from another branch that is not from my fork. Adaptive scheme for stochastic Runge-Kutta methods due to Roessler #182 .

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -78,8 +78,17 @@ const is_APPVEYOR = Sys.iswindows() && haskey(ENV,"APPVEYOR")
   if !is_APPVEYOR && (GROUP == "All" || GROUP == "WeakConvergence3")
     @time @safetestset "Roessler weak SRK Tests" begin include("weak_convergence/srk_weak_final.jl") end
   end
-
+  
   if !is_APPVEYOR && (GROUP == "All" || GROUP == "WeakConvergence4")
       @time @safetestset "Weak Stratonovich Tests" begin include("weak_convergence/weak_strat.jl") end
   end
+
+  if !is_APPVEYOR && GROUP == "WeakConvergence"
```
Comment (ChrisRackauckas) [inline test/runtests.jl:86]:
```suggestion
  if !is_APPVEYOR && GROUP == "WeakAdaptive"
```

? Also, need to add the Gitlab group

**Comment by frankschae** [comment]
Instead of using only the final weak error, I am comparing the expected value at several time points. Without the previous bug in the solve call, the adaptive scheme looks good (see Plots). 

adaptive-1 0
adaptive-0 1
adaptive-0 01

**Comment by frankschae** [comment]
@ChrisRackauckas could you take a look on the tests? I am a bit puzzled.. it still says:
"459 - Run `import Pkg; Pkg.add("Pkg")` to install the Pkg package." but this points to the line in the runscript where I have already added the` import Pkg` command...

**Comment by frankschae** [comment]
@YingboMa  , @ChrisRackauckas 

Do you have some suggestions how to fix the generation of the three-point distributed random variables? Adding inbounds makes the CPU version faster but gives the following error:

InvalidIRError: compiling kernel calc_threepoint_random2!(CuDeviceArray{Float32,1,CUDA.AS.Global}, Float32, Float32, CuDeviceArray{Float32,1,CUDA.AS.Global}) resulted in invalid LLVM IR
Reason: unsupported dynamic function invocation (call to print)
Stacktrace:
 [1] print_to_string at strings/io.jl:135
 [2] string at strings/io.jl:174
 [3] throw_complex_domainerror at math.jl:33
 [4] multiple call sites at unknown:0
Stacktrace:
 [1] check_ir(::GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget,CUDA.CUDACompilerParams}, ::LLVM.Module) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/validation.jl:123
 [2] macro expansion at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/driver.jl:187 [inlined]
 [3] macro expansion at /home/frank/.julia/packages/TimerOutputs/dVnaw/src/TimerOutput.jl:206 [inlined]
 [4] codegen(::Symbol, ::GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget,CUDA.CUDACompilerParams}; libraries::Bool, deferred_codegen::Bool, optimize::Bool, strip::Bool, strict::Bool) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/driver.jl:185
 [5] compile(::Symbol, ::GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget,CUDA.CUDACompilerParams}; libraries::Bool, deferred_codegen::Bool, optimize::Bool, strip::Bool, strict::Bool) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/driver.jl:36
 [6] _cufunction(::GPUCompiler.FunctionSpec{typeof(calc_threepoint_random2!),Tuple{CuDeviceArray{Float32,1,CUDA.AS.Global},Float32,Float32,CuDeviceArray{Float32,1,CUDA.AS.Global}}}; kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /home/frank/.julia/packages/CUDA/5t6R9/src/compiler/execution.jl:308
 [7] _cufunction at /home/frank/.julia/packages/CUDA/5t6R9/src/compiler/execution.jl:302 [inlined]
 [8] #75 at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/cache.jl:21 [inlined]
 [9] get!(::GPUCompiler.var"#75#76"{Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},typeof(CUDA._cufunction),GPUCompiler.FunctionSpec{typeof(calc_threepoint_random2!),Tuple{CuDeviceArray{Float32,1,CUDA.AS.Global},Float32,Float32,CuDeviceArray{Float32,1,CUDA.AS.Global}}}}, ::Dict{UInt64,Any}, ::UInt64) at ./dict.jl:452
 [10] macro expansion at ./lock.jl:183 [inlined]
 [11] check_cache(::typeof(CUDA._cufunction), ::GPUCompiler.FunctionSpec{typeof(calc_threepoint_random2!),Tuple{CuDeviceArray{Float32,1,CUDA.AS.Global},Float32,Float32,CuDeviceArray{Float32,1,CUDA.AS.Global}}}, ::UInt64; kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/cache.jl:19
 [12] + at ./int.jl:53 [inlined]
 [13] hash_64_64 at ./hashing.jl:35 [inlined]
 [14] hash_uint64 at ./hashing.jl:62 [inlined]
 [15] hx at ./float.jl:568 [inlined]
 [16] hash at ./float.jl:571 [inlined]
 [17] cached_compilation(::typeof(CUDA._cufunction), ::GPUCompiler.FunctionSpec{typeof(calc_threepoint_random2!),Tuple{CuDeviceArray{Float32,1,CUDA.AS.Global},Float32,Float32,CuDeviceArray{Float32,1,CUDA.AS.Global}}}, ::UInt64; kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/cache.jl:0
 [18] cached_compilation(::Function, ::GPUCompiler.FunctionSpec{typeof(calc_threepoint_random2!),Tuple{CuDeviceArray{Float32,1,CUDA.AS.Global},Float32,Float32,CuDeviceArray{Float32,1,CUDA.AS.Global}}}, ::UInt64) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/cache.jl:37
 [19] cufunction(::Function, ::Type; name::Nothing, kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /home/frank/.julia/packages/CUDA/5t6R9/src/compiler/execution.jl:296
 [20] cufunction(::Function, ::Type{T} where T) at /home/frank/.julia/packages/CUDA/5t6R9/src/compiler/execution.jl:291
 [21] top-level scope at /home/frank/.julia/packages/CUDA/5t6R9/src/compiler/execution.jl:108

Here is a MWE:

```julia
using BenchmarkTools, Random, Test

function calc_threepoint_random2!(_dW, dt, quantile, dW_scaled)
    for i = 1:length(dW_scaled)
        @inbounds begin
          if dW_scaled[i] < quantile
            _dW[i] = -sqrt(3*dt)
          elseif dW_scaled[i] > -quantile
            _dW[i] = sqrt(3*dt)
          else
            _dW[i] = zero(dt)
          end
        end
    end
    return nothing
end

seed = 100
Random.seed!(seed)
chi1 = randn(Float32,100000)
_dW = zero(chi1)

dt = 1.0f0
quantile = -0.9674215661017014f0

@btime calc_threepoint_random2!(_dW2, dt, quantile, chi1)
@cuda calc_threepoint_random2!(_dWGPU, dt, quantile, chiGPU)
``` 

Also for the two-point distributed variables the control flow statements seem to cause problems:
``` julia
function calc_twopoint_random!(_dW, dt, dW)
    @. _dW = ifelse(sign(dW) > 0.0, dt, -dt)
    return nothing
end

@cuda add_2!(_dWGPU, chiGPU)
```

ERROR: GPU compilation of kernel add_2!(CuDeviceArray{Float32,1,CUDA.AS.Global}, CuDeviceArray{Float32,1,CUDA.AS.Global}) failed
KernelError: recursion is currently not supported

Try inspecting the generated code with any of the @device_code_... macros.

Stacktrace:
 [1] _nextind_str at strings/string.jl:141
 [2] nextind at strings/string.jl:137
 [3] _nextind_str at strings/string.jl:141
 [4] match at regex.jl:275
 [5] _splitdir_nodrive at path.jl:138
 [6] __configure__ at /home/frank/.julia/packages/CUDA/5t6R9/src/initialization.jl:85
 [7] #673 at /home/frank/.julia/packages/CUDA/5t6R9/src/initialization.jl:27
 [8] lock at lock.jl:159
 [9] _functional at /home/frank/.julia/packages/CUDA/5t6R9/src/initialization.jl:26
 [10] functional at /home/frank/.julia/packages/CUDA/5t6R9/src/initialization.jl:18
 [11] #1506#cache_fptr! at /home/frank/.julia/packages/CUDA/5t6R9/lib/utils/call.jl:31
 [12] cuEventCreate at /home/frank/.julia/packages/CUDA/5t6R9/lib/utils/call.jl:93
 [13] CuEvent at /home/frank/.julia/packages/CUDA/5t6R9/lib/cuda/events.jl:18
 [14] add_2! at /home/frank/ownCloud/Private/GSoC/SDE_new_algos/adaptive.jl:691
Stacktrace:
 [1] push!(::GPUCompiler.MethodCompileTracer, ::Core.MethodInstance) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/irgen.jl:52
 [2] (::GPUCompiler.var"#hook_emit_function#24"{GPUCompiler.MethodCompileTracer})(::Core.MethodInstance, ::Core.CodeInfo, ::UInt64) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/irgen.jl:210
 [3] compile_method_instance(::GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget,CUDA.CUDACompilerParams}, ::Core.MethodInstance, ::UInt64) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/irgen.jl:274
 [4] macro expansion at /home/frank/.julia/packages/TimerOutputs/dVnaw/src/TimerOutput.jl:206 [inlined]
 [5] irgen(::GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget,CUDA.CUDACompilerParams}, ::Core.MethodInstance, ::UInt64) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/irgen.jl:332
 [6] macro expansion at /home/frank/.julia/packages/TimerOutputs/dVnaw/src/TimerOutput.jl:206 [inlined]
 [7] macro expansion at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/driver.jl:98 [inlined]
 [8] macro expansion at /home/frank/.julia/packages/TimerOutputs/dVnaw/src/TimerOutput.jl:206 [inlined]
 [9] codegen(::Symbol, ::GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget,CUDA.CUDACompilerParams}; libraries::Bool, deferred_codegen::Bool, optimize::Bool, strip::Bool, strict::Bool) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/driver.jl:97
 [10] compile(::Symbol, ::GPUCompiler.CompilerJob{GPUCompiler.PTXCompilerTarget,CUDA.CUDACompilerParams}; libraries::Bool, deferred_codegen::Bool, optimize::Bool, strip::Bool, strict::Bool) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/driver.jl:36
 [11] _cufunction(::GPUCompiler.FunctionSpec{typeof(add_2!),Tuple{CuDeviceArray{Float32,1,CUDA.AS.Global},CuDeviceArray{Float32,1,CUDA.AS.Global}}}; kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /home/frank/.julia/packages/CUDA/5t6R9/src/compiler/execution.jl:308
 [12] _cufunction at /home/frank/.julia/packages/CUDA/5t6R9/src/compiler/execution.jl:302 [inlined]
 [13] #75 at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/cache.jl:21 [inlined]
 [14] get!(::GPUCompiler.var"#75#76"{Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},typeof(CUDA._cufunction),GPUCompiler.FunctionSpec{typeof(add_2!),Tuple{CuDeviceArray{Float32,1,CUDA.AS.Global},CuDeviceArray{Float32,1,CUDA.AS.Global}}}}, ::Dict{UInt64,Any}, ::UInt64) at ./dict.jl:452
 [15] macro expansion at ./lock.jl:183 [inlined]
 [16] check_cache(::typeof(CUDA._cufunction), ::GPUCompiler.FunctionSpec{typeof(add_2!),Tuple{CuDeviceArray{Float32,1,CUDA.AS.Global},CuDeviceArray{Float32,1,CUDA.AS.Global}}}, ::UInt64; kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/cache.jl:19
 [17] + at ./int.jl:53 [inlined]
 [18] hash_64_64 at ./hashing.jl:35 [inlined]
 [19] hash_uint64 at ./hashing.jl:62 [inlined]
 [20] hx at ./float.jl:568 [inlined]
 [21] hash at ./float.jl:571 [inlined]
 [22] cached_compilation(::typeof(CUDA._cufunction), ::GPUCompiler.FunctionSpec{typeof(add_2!),Tuple{CuDeviceArray{Float32,1,CUDA.AS.Global},CuDeviceArray{Float32,1,CUDA.AS.Global}}}, ::UInt64; kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/cache.jl:0
 [23] cached_compilation(::Function, ::GPUCompiler.FunctionSpec{typeof(add_2!),Tuple{CuDeviceArray{Float32,1,CUDA.AS.Global},CuDeviceArray{Float32,1,CUDA.AS.Global}}}, ::UInt64) at /home/frank/.julia/packages/GPUCompiler/lqbF2/src/cache.jl:37
 [24] cufunction(::Function, ::Type; name::Nothing, kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /home/frank/.julia/packages/CUDA/5t6R9/src/compiler/execution.jl:296
 [25] cufunction(::Function, ::Type{T} where T) at /home/frank/.julia/packages/CUDA/5t6R9/src/compiler/execution.jl:291
 [26] top-level scope at /home/frank/.julia/packages/CUDA/5t6R9/src/compiler/execution.jl:108


The generation of the three-point random variables should be the last issue before a simplified version of the DRI1 method (for diagonal and non-mixing noise) can be run on a GPU (without scalar operations).

**Comment by ChrisRackauckas** [comment]
```julia
@. _dW = (sign(dW) > 0.0)*2dt -dt
```

**Comment by frankschae** [comment]
Does this work for you? I still get the same " recursion is currently not supported" message

```julia
using CUDA, Random

seed = 100
Random.seed!(seed)
chi1 = randn(Float32,100000)
_dW = zero(chi1)
chiGPU = CuArray(chi1)
_dWGPU =  zero(chiGPU)

dt = 1.0f0

function calc_twopoint_random2!(_dW, dt, dW)
  @. _dW = (sign(dW) > zero(typeof(dW)))*2*dt-dt
  return nothing
end

@cuda calc_twopoint_random2!(_dWGPU, dt, chiGPU)
```

**Comment by ChrisRackauckas** [comment]
Haven't been able to test yet. Try making it a bit simpler: `@. _dW = (sign(dW) > false)*2*dt-dt`?

**Comment by frankschae** [comment]
oh..
```
 @. _dW = dW
```
also fails.. but
```
 @. _dW = dt
```
compiles.

**Comment by ChrisRackauckas** [comment]
that sounds like a bug. I'd file something on that.

**Comment by frankschae** [comment]
I just compared with respect to their `add_broadcast!()` https://juliagpu.gitlab.io/CUDA.jl/tutorials/introduction/
This works:
``` julia

using CUDA, Random, BenchmarkTools

seed = 100
Random.seed!(seed)
chi1 = randn(Float32,100000)
_dW = zero(chi1)
chiGPU = CuArray(chi1)
_dWGPU =  zero(chiGPU)

dt = 1.0f0

function calc_twopoint_random2!(_dW, dt, dW)
  CUDA.@sync @. _dW = (sign(dW) > zero(typeof(dW)))*2*dt-dt
  return nothing
end

@btime calc_twopoint_random2!(_dWGPU, dt, chiGPU)
```
I thought that I'd need `@cuda` to launch the kernel.

**Comment by frankschae** [comment]
yep, I also just found that the initial version 
```
function calc_twopoint_random2!(_dW, dt, dW)
  @. _dW = ifelse(sign(dW) > 0.0, dt, -dt)
  return nothing
end

@btime calc_twopoint_random2!(_dWGPU, dt, chiGPU)
```
just works fine and is even a bit faster. So what is the @cuda good for?

**Comment by ChrisRackauckas** [comment]
writing your own kernels

**Comment by frankschae** [comment]
the diff somehow shows too much. It contains the changes for `PL1WM` which were merged earlier..

**Comment by ChrisRackauckas** [comment]
probably an issue with the merge. Can you cherrypick to a new PR?

**Comment by ChrisRackauckas** [comment]
GPU CI got a new image and should be up and running now

**Comment by frankschae** [comment]
https://travis-ci.com/github/SciML/StochasticDiffEq.jl/jobs/359540217

Looks like I accidentally changed something for tau leaping:

Basic Tau Leaping Tests: Test Failed at /home/travis/build/SciML/StochasticDiffEq.jl/test/tau_leaping.jl:46
746  Expression: ≈(mean1, meanX, rtol = 0.01)
747   Evaluated: 721.9698 ≈ 729.8305 (rtol=0.01)

**Comment by ChrisRackauckas** [comment]
that is most likely stochastic. Let me check though.

**Comment by ChrisRackauckas** [comment]
Use CUDA.jl instead of CuArrays.jl now, and https://travis-ci.com/github/SciML/StochasticDiffEq.jl/jobs/359540216#L812 it seems like there's real test errors.

**Comment by ChrisRackauckas** [comment]
>Looks like I accidentally changed something for tau leaping:

That one's a stochastic failure. It has a low chance of failing, but does periodically. Maybe we should make the trajectories a little higher, but I can investigate that later if it happens too often.

**Comment by frankschae** [comment]
oh yeah I forgot to replace it for the Brusselator example where only `EnsembleCPUArray()` is used

**Comment by ChrisRackauckas** [comment]
@christopher-dG could the execution limits be increased here? https://gitlab.com/JuliaGPU/StochasticDiffEq.jl/-/jobs/633913058

@frankschae looks like that weak convergence test file is... massive 🤣 . It might need to be split up. It would be better to run it on GPUs of course, but let's do this one step at a time: get it on CPUs, then start GPUing it. We can get some help getting that thing running in a way that it can handle the big tests.

Right now the issue is that the test queue is mixing in some bad GPU machines with our few good ones, so we need to make sure we bump the GPU memory limit up and require something good, since these tests will be like 20 minute big tests that will require their own GPU. Pinging @maleadt so he's aware. 

If this requires new hardware let me know, but some of the motley crew of random machines should be able to handle this, so it really just needs to remove the limits.

**Comment by christopher-dG** [comment]
I bumped the timeout to 12h.

**Comment by ChrisRackauckas** [comment]
Awesome. Let's destroy that machine.

**Comment by frankschae** [comment]
yeah.. it contains quite a lot since it tests all noise forms for all the new SRKs. It ran until the very last one though (so if starting times with Julia 1.5 are decreased I guess it'd work already now but it's probably anyways better to split one expensive noise test off.)

Regarding speed of the weak tests in general. Would it be worth to look at the variance reduction methods described in Kloeden & Platen's book?

**Comment by ChrisRackauckas** [comment]
>yeah.. it contains quite a lot since it tests all noise forms for all the new SRKs. It ran until the very last one though (so if starting times with Julia 1.5 are decreased I guess it'd work already now but it's probably anyways better to split one expensive noise test off.)

Yeah, it should be the worst of times. v1.6 has a change that will cause massive improvements to small SDEs (because right now there's some dynamic behavior due to keyword arguments), so this should definitely be gone fairly soon. That said...

>Regarding speed of the weak tests in general. Would it be worth to look at the variance reduction methods described in Kloeden & Platen's book?

Yes, we could try a few of those. Also, I think it would be very cool to build a new library around multilevel Monte Carlo methods. These methods are too memory heavy for neural SDEs since they work on all time points at once, and cannot support things like events (which is why they'd need a separate repo), but it would be really cool to build a library around them:

https://people.maths.ox.ac.uk/gilesm/files/OPRE_2008.pdf
https://arxiv.org/pdf/1202.6283.pdf
https://people.maths.ox.ac.uk/gilesm/files/mcqmc14.pdf
https://arxiv.org/pdf/1412.4069.pdf

**Comment by frankschae** [comment]
Do you know this package https://github.com/PieterjanRobbe/MultilevelEstimators.jl ?
Looks like at least the scheme https://people.maths.ox.ac.uk/gilesm/files/OPRE_2008.pdf exists already.

**Comment by ChrisRackauckas** [comment]
Oh I hadn't seen that. It doesn't look parallelized though and I don't see the easy way to make it act on SDEProblem, so there's still some work to be done to munge it towards something I'd like to see. But that's definitely one direction!

**Comment by ChrisRackauckas** [comment]
Was that last rebase supposed to include a test split?

**Comment by frankschae** [comment]
Alright, cool. I'd definitely be happy to contribute when we know where to start with a compatible MLMC package.

No, I just resolved the merge conflict -- I'll split the tests later today. First I'd like to finish the MTK PR.

**Comment by ChrisRackauckas** [comment]
Looks like the GPU tests are almost working. I don't want to have that hold back this merge, but it's something to look into, even maybe just a paired down version on GPU CI

**Pull request merged**